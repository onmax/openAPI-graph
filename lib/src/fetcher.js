"use strict";
/**
 * The main goal is to find all openAPI specification given a path.
 * Therefore, it will loop over all files and creating a structured object
 * as a result
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.testables = exports.fetcher = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const openapi_1 = require("./openapi");
async function fetcher(path) {
    // Converts path to absolute
    path = path_1.resolve(path);
    const pathExists = fs_1.existsSync(path);
    if (!pathExists) {
        throw new Error('The given path does not exist in your system');
    }
    else {
        return await loadsSwaggerFiles(path);
    }
}
exports.fetcher = fetcher;
/**
 * Finds all the valid OpenApi specifications given a directory and returns a
 * list of all the specification contents
 *
 * @param projectPath of the project where the specifications are suppose to be
 * @returns a list of OpenAPIContent which has the path, openAPI version
 *          and the content for every specification found inside of the given path
 */
async function loadsSwaggerFiles(projectPath) {
    const projectContent = fs_1.readdirSync(projectPath);
    if (projectContent === undefined || projectContent.length === 0) {
        return [];
    }
    const paths = projectContent
        // Find all yml and yaml files
        .filter((p) => p.match(/.*\.(yml|yaml)/gi))
        .map(p => `${projectPath}/${p}`);
    return await openapi_1.getOpenApisContent(paths);
}
// Just for testing reasons https://stackoverflow.com/a/54116079
exports.testables = {
    loadsSwaggerFiles,
};
