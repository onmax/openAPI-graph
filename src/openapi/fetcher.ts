/**
 * The main goal is to find all openAPI specification given a path.
 * Therefore, it will loop over all files and creating a structured object
 * as a result
 *
 */

import { existsSync, readdirSync } from 'fs';
import { resolve } from 'path';
import { OpenAPIContent } from 'openapi-graph-types';
import { getOpenApisContent } from '.';
const logger = require('pino')({
  prettyPrint: {
    ignore: 'time,pid,hostname',
    singleLine: true
  }
})

const COMMON_FOLDER_NAMES = ["node_modules", "target"]
const COMMON_FILES_NAMES = ["docker-compose"].flatMap(e => [`${e}.yml`, `${e}.yaml`])

export async function fetcher(paths: string[]): Promise<OpenAPIContent[]>;
export async function fetcher(path: string): Promise<OpenAPIContent[]>;
export async function fetcher(input: any): Promise<OpenAPIContent[]> {
  if (typeof input === 'string') {
    // If we are working with a path, then it means that we have to find all openAPI specifications

    // Converts path to absolute
    const resolvedPath = resolve(input);
    const pathExists = existsSync(resolvedPath);
    if (!pathExists) {
      throw new Error('The given path does not exist in your system');
    } else {
      return await loadsSwaggerFiles(resolvedPath);
    }
  } else if (([]).constructor === input.constructor) {
    // If we are working with a path, then it means that we have to find all openAPI specifications
    const resolvedPaths = (input as string[])
      .map(i => resolve(input))
      .filter(resolvedPath => existsSync(resolvedPath) || logger.console.warn(`Ignoring ${resolvedPath} because the path is not reachable`));
    return getOpenApisContent(resolvedPaths);
  } else {
    logger.console.warn(`Invalid given input. It was expected a string or array of strings. Received ${input}.`);
    return [];
  }

}

/**
 * Finds all the valid OpenApi specifications given a directory and returns a
 * list of all the specification contents
 *
 * @param projectPath of the project where the specifications are suppose to be
 * @returns a list of OpenAPIContent which has the path, openAPI version
 *          and the content for every specification found inside of the given path
 */
async function loadsSwaggerFiles(projectPath: string): Promise<OpenAPIContent[]> {
  const projectContent = await getFiles(resolve(projectPath));
  logger.info(`Found (${projectContent.length}) ${projectContent}`);
  if (projectContent === undefined || projectContent.length === 0) {
    return [];
  }

  return await getOpenApisContent(projectContent);
}

async function getFiles(fromPath = './', paths: string[] = []): Promise<string[]> {
  const entries = await readdirSync(fromPath, { withFileTypes: true });
  for (const entry of entries) {
    // We don't check hidden files or folders
    if (!/(^|\/)\.[^/.]/g.test(entry.name)) {
      if (
        entry.isDirectory() &&
        // Don't check common folder names
        !COMMON_FOLDER_NAMES.includes(entry.name)) {
        paths.push(...(await getFiles(`${fromPath}/${entry.name}/`)));
      } else if (
        entry.isFile() &&
        // JSON are also valid, but they are more generic, so we don't look for them because it takes a lot of time
        entry.name.match(/.*\.(yml|yaml)/gi) &&
        // Don't check common file names
        !COMMON_FILES_NAMES.includes(entry.name)) {
        paths.push(resolve(`${fromPath}/${entry.name}`));
      }
    }
  }
  return paths;
}

// Just for testing reasons https://stackoverflow.com/a/54116079
export const testables = {
  loadsSwaggerFiles,
  getFiles,
};
