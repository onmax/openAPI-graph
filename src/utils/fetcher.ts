/**
 * The main goal is to find all openAPI specification given a path.
 * Therefore, it will loop over all files and creating a structured object
 * as a result
 *
 */

import { existsSync, readdirSync, readFileSync } from 'fs';
import { load as loadYaml } from 'js-yaml';
import { LogLevel, OpenAPIContent } from 'openapi-graph-types';
import { resolve } from 'path';
import { log } from '.';

const COMMON_FOLDER_NAMES = ["node_modules", "target"]
const COMMON_FILES_NAMES = ["docker-compose"].flatMap(e => [`${e}.yml`, `${e}.yaml`])

export async function fetcher(paths: string[]): Promise<OpenAPIContent[]>;
export async function fetcher(path: string): Promise<OpenAPIContent[]>;
export async function fetcher(input: any): Promise<OpenAPIContent[]> {
  if (typeof input === 'string') {
    // If we are working with a path, then it means that we have to find all openAPI specifications

    // Converts path to absolute
    const resolvedPath = resolve(input);
    const pathExists = existsSync(resolvedPath);
    if (!pathExists) {
      throw new Error('The given path does not exist in your system');
    } else {
      return await loadsSwaggerFiles(resolvedPath);
    }
  } else if (([]).constructor === input.constructor) {
    // If we are working with a path, then it means that we have to find all openAPI specifications
    const resolvedPaths = (input as string[])
      .map(i => resolve(input))
      .filter(resolvedPath => existsSync(resolvedPath) || log(`Ignoring ${resolvedPath} because the path is not reachable`, LogLevel.WARN));
    return getOpenApisContent(resolvedPaths);
  } else {
    log(`Invalid given input. It was expected a string or array of strings. Received ${input}`, LogLevel.WARN);
    return [];
  }

}

/**
 * Finds all the valid OpenApi specifications given a directory and returns a
 * list of all the specification contents
 *
 * @param projectPath of the project where the specifications are suppose to be
 * @returns a list of OpenAPIContent which has the path, openAPI version
 *          and the content for every specification found inside of the given path
 */
async function loadsSwaggerFiles(projectPath: string): Promise<OpenAPIContent[]> {
  const projectContent = await getFiles(resolve(projectPath));
  log(`Found (${projectContent.length}) ${projectContent}`);
  if (projectContent === undefined || projectContent.length === 0) {
    return [];
  }

  return await getOpenApisContent(projectContent);
}

async function getFiles(fromPath = './', paths: string[] = []): Promise<string[]> {
  const entries = await readdirSync(fromPath, { withFileTypes: true });
  for (const entry of entries) {
    // We don't check hidden files or folders
    if (!/(^|\/)\.[^/.]/g.test(entry.name)) {
      if (
        entry.isDirectory() &&
        // Don't check common folder names
        !COMMON_FOLDER_NAMES.includes(entry.name)) {
        paths.push(...(await getFiles(`${fromPath}/${entry.name}/`)));
      } else if (
        entry.isFile() &&
        // JSON are also valid, but they are more generic, so we don't look for them because it takes a lot of time
        entry.name.match(/.*\.(yml|yaml)/gi) &&
        // Don't check common file names
        !COMMON_FILES_NAMES.includes(entry.name)) {
        paths.push(resolve(`${fromPath}/${entry.name}`));
      }
    }
  }
  return paths;
}

/**
 * reads the contents of all given paths. The files will be validated.
 *
 * @param paths of the .yaml|.yml files. The paths should exists.
 * @returns the contents of the files as JSON with their path
 */
async function getOpenApisContent(paths: string[]): Promise<OpenAPIContent[]> {
  return (
    paths
      .map(p => ({ path: p, content: loadYaml(readFileSync(p, 'utf8')) || log(`Couldn't load ${p}`, LogLevel.WARN) }))
      .filter(p => p.content?.openapi?.includes("3.0"))
  );
}

// Just for testing reasons https://stackoverflow.com/a/54116079
export const testables = {
  loadsSwaggerFiles,
  getFiles,
  getOpenApisContent
};
