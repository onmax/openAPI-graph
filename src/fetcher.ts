/**
 * The main goal is to find all openAPI specification given a path.
 * Therefore, it will loop over all files and creating a structured object
 * as a result
 *
 */

import { existsSync, readdirSync } from 'fs';
import { resolve } from 'path';
import { OpenAPIContent } from '../model';
import { getOpenApisContent } from './openapi';

export async function fetcher(path: string): Promise<OpenAPIContent[]> {
  // Converts path to absolute
  path = resolve(path);
  const pathExists = existsSync(path);

  if (!pathExists) {
    throw new Error('The given path does not exist in your system');
  } else {
    return await loadsSwaggerFiles(path);
  }
}

/**
 * Finds all the valid OpenApi specifications given a directory and returns a
 * list of all the specification contents
 *
 * @param projectPath of the project where the specifications are suppose to be
 * @returns a list of OpenAPIContent which has the path, openAPI version
 *          and the content for every specification found inside of the given path
 */
async function loadsSwaggerFiles(projectPath: string): Promise<OpenAPIContent[]> {
  const projectContent = await getFiles(resolve(projectPath));
  if (projectContent === undefined || projectContent.length === 0) {
    return [];
  }

  return await getOpenApisContent(projectContent);
}

async function getFiles(fromPath = './', paths: string[] = []): Promise<string[]> {
  const entries = await readdirSync(fromPath, { withFileTypes: true });
  for (const entry of entries) {
    if (entry.isDirectory()) {
      paths.push(...(await getFiles(`${fromPath}/${entry.name}/`)));
    } else if (entry.name.match(/.*\.(yml|yaml|json)/gi)) {
      paths.push(resolve(`${fromPath}/${entry.name}`));
    }
  }
  return paths;
}

// Just for testing reasons https://stackoverflow.com/a/54116079
export const testables = {
  loadsSwaggerFiles,
  getFiles,
};
