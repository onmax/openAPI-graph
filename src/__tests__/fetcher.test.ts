import { resolve } from 'path';

import { fetcher, testables } from '../fetcher';
const { loadsSwaggerFiles } = testables;

const resourcesDir = `${resolve('.')}/src/__tests__/resources`;
const noneProject = `${resourcesDir}/none`;
const petStoreProject = `${resourcesDir}/petstore`;

test('Should not load anything', async () => {
    expect(await loadsSwaggerFiles(noneProject)).toEqual([]);
});

test('Should retrieve all the valid openAPI files and load them as JSON in the petstore folder', async () => {
    (await loadsSwaggerFiles(petStoreProject)).forEach((o) => {
        expect(o).toMatchSnapshot({
            path: expect.any(String),
            content: expect.any(Object),
        });
        expect(o.content).toMatchSnapshot({
            openapi: expect.any(String),
            info: expect.any(Object),
            paths: expect.any(Object),
        });
    });
});

test('Should raise exception because path does not exists', () => {
    expect(fetcher("this/path/does/not/exists")).rejects.toStrictEqual(new Error('The given path does not exist in your system'));
});
