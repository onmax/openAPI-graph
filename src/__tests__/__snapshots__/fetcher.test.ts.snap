// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should retrieve all the valid openAPI files and load them as JSON in the petstore folder 1`] = `
Object {
  "content": Any<Object>,
  "path": Any<String>,
}
`;

exports[`Should retrieve all the valid openAPI files and load them as JSON in the petstore folder 2`] = `
Object {
  "components": Object {
    "schemas": Object {
      "Error": Object {
        "properties": Object {
          "code": Object {
            "format": "int32",
            "type": "integer",
          },
          "message": Object {
            "type": "string",
          },
        },
        "required": Array [
          "code",
          "message",
        ],
        "type": "object",
      },
      "Pet": Object {
        "properties": Object {
          "id": Object {
            "format": "int64",
            "type": "integer",
          },
          "name": Object {
            "type": "string",
          },
          "tag": Object {
            "type": "string",
          },
        },
        "required": Array [
          "id",
          "name",
        ],
        "type": "object",
      },
      "Pets": Object {
        "items": Object {
          "$ref": "#/components/schemas/Pet",
        },
        "type": "array",
      },
      "SchemaNotBeingUsed": Object {
        "properties": Object {
          "key": Object {
            "type": "string",
          },
          "value": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
    },
  },
  "info": Any<Object>,
  "openapi": Any<String>,
  "paths": Any<Object>,
  "servers": Array [
    Object {
      "url": "http://petstore.swagger.io/v1",
    },
  ],
}
`;
